apis/v1alpha1/observabilitypolicy_types.go

```diff
@@ -0,0 +1,124 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	gatewayv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:categories=nginx-gateway-fabric,scope=Namespaced
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+// +kubebuilder:metadata:labels="gateway.networking.k8s.io/policy=direct"
+
+// ObservabilityPolicy is a Direct Attached Policy. It provides a way to configure observability settings for
+// the NGINX Gateway Fabric data plane. Used in conjunction with the NginxProxy CRD that is attached to the
+// GatewayClass parametersRef.
+type ObservabilityPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the ObservabilityPolicy.
+	Spec ObservabilityPolicySpec `json:"spec"`
+
+	// Status defines the state of the ObservabilityPolicy.
+	Status gatewayv1alpha2.PolicyStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ObservabilityPolicyList contains a list of ObservabilityPolicies.
+type ObservabilityPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ObservabilityPolicy `json:"items"`
+}
+
+// ObservabilityPolicySpec defines the desired state of the ObservabilityPolicy.
+type ObservabilityPolicySpec struct {
+	// TargetRef identifies an API object to apply the policy to.
+	// Object must be in the same namespace as the policy.
+	//
+	// Support: HTTPRoute
+	TargetRef gatewayv1alpha2.PolicyTargetReference `json:"targetRef"`
+
+	// Tracing allows for enabling and configuring tracing.
+	//
+	// +optional
+	Tracing *Tracing `json:"tracing,omitempty"`
+}
+
+// Tracing allows for enabling and configuring OpenTelemetry tracing.
+type Tracing struct {
+	// Enable defines if tracing is enabled, disabled, or parent-based.
+	Enable TraceType `json:"enable"`
```

Comment:
I wonder if the allowed values `on` and `off` could be source of annoyance for our users here, because YAML treat those values are bools.

For example:
```yaml
  tracing:
    enable: on
```
```
kubectl apply -f example.yaml
The ObservabilityPolicy "example-observabilitypolicy" is invalid:
* spec.tracing.enable: Invalid value: "boolean": spec.tracing.enable in body must be of type string: "boolean"
* spec.tracing.enable: Unsupported value: true: supported values: "on", "off", "parent"
```

So user will need to figure it out to put the value in `"..."`.

```yaml
  tracing:
    enable: "on"
```

```
kubectl apply -f example.yaml
observabilitypolicy.gateway.nginx.org/example-observabilitypolicy created
```

Reply:
Trying to think of alternatives...`basic`, `none`, `parent`? Should we rename the field from `enable` to something else as well to make sense with the possible values?

Reply:
What if we have enable as bool and have an extra field that will work for ration and for parent

Something like that:
```yaml
enable: true
precondition: "ratio"
ratio: 15 # only allowed if precondition is "ratio"
```

```yaml
enable: true
precondition: "parent"
```

Something more verbose, but potentially extendable to other cases:

```yaml
enable: true
preconditions:
- type: Ratio
  ratioPrecondition:
    ratio: 15
```

```yaml
enable: true
preconditions:
- type: Parent
  parentPrecondition: {} # it doesn't need any values though
```

Reply:
Hm, feels a bit complex for a pretty simple config. Ratio-based is technically the default, so "parent" is really the outlier here.

Reply:
The former option seems a bit more doable I think, though.

Reply:
after more thinking, I wonder if we need to have "enable/disable" (on/off) switch at all.

what if we treat the policy as always enabled?

and for extra tuning, what if we just had:
```yaml
ratio: 100
. . .
```
```yaml
parent: true
. . .
```

and those fields are probably mutually exclusive?

Reply:
> I think here we might create a precedent, which we will need to follow later.

That's a good point. I guess my only thoughts were that it would be nice for a user to be able to fill out a tracing config and have an easy way to turn it on/off without having to repopulate their whole config, but is that pattern really used anywhere else?

Reply:
I guess once we have also logging and/or metric there, disabling only tracing with those options above will not be possible

Reply:
Though this goes back to "I want to keep metrics enabled, but temporarily disable tracing". This is probably overthinking and a less common use case.

targetRef is a required field

Reply:
Just to add to the conversation, here's what the k8s API conventions doc has to say about booleans:

> Think twice about bool fields. Many ideas start as boolean but eventually trend towards a small set of mutually exclusive options. Plan for future expansions by describing the policy options explicitly as a string type alias (e.g. TerminationMessagePolicy).

I think we should either remove the `enable` field and assume non-nil tracing config means enabled,  or go with Saylor's first alternative of changing the field name and enum values. Maybe `type` or `strategy` with [`simple`, `parent`, `ratio`, `none`]

Reply:
I like `strategy` more than `type`. We wouldn't need `simple`, since `simple` is just a ratio of 100%. So the question is do we want a `none` option or just set tracing to nil...

Reply:
Can we add a `none` option at a later date without a breaking change?

Reply:
> Can we add a none option at a later date without a breaking change?

Yes

Reply:
If we set validation to have `ratio: x` and `strategy: parent` to be mutually exclusive, then we can't set a ratio default.